### PR TITLE
[DOC release] Add note about router service event and property update…

### PR DESCRIPTION
**Pull request by xg-wang**
… timing

Also adding a bunch of hook/event assertions in tests to clearly state the supported timing order.

Related https://github.com/emberjs/ember.js/issues/18665

### CODE REVIEW

**Comment by xg-wang** [comment]
I'm not entirely sure if it should be documented as public for exact timing order. In fact I would like to deprecate `Route`'s `didTransition` action, then make it public saying `currentURL` etc updates after model hooks and before the final `routeDidChange` event during mid-transition.

File: packages/@ember/-internals/routing/lib/services/router.ts [hunk 1]
```diff
@@ -439,6 +439,10 @@ export default class RouterService extends Service {
     }
     ```
 
+    The exact timing to fire `routeDidChange` is after the Route's
+    [didTransition](/ember/release/classes/Route/events/didTransition?anchor=didTransition)
+    action fired.
```
Comment (rwjblue) [inline packages/@ember/-internals/routing/lib/services/router.ts]: I think that it is good to explicitly state this, but I'd prefer to tweak the verbiage here a bit. Maybe something:

```
`routeDidChange` will be called after any `Route`'s [didTransition](/ember/release/classes/Route/events/didTransition?anchor=didTransition) action has been fired.
```

File: packages/@ember/-internals/routing/lib/services/router.ts [hunk 2]
```diff
@@ -469,6 +473,9 @@ RouterService.reopen(Evented, {
      * `blog.index` when you visit `/blog`
      * `blog.post` when you visit `/blog/some-post-id`
 
+     This property is updated after Route's [didTransition](/ember/release/classes/Route/events/didTransition?anchor=didTransition)
+     action and before `routeDidChange` being fired.
```
Comment (rwjblue) [inline packages/@ember/-internals/routing/lib/services/router.ts]: FWIW, I would actually prefer to change these to actually be based on `routeDidChange`.

**Review by rwjblue — COMMENTED** [review]
Overall, I think this is great! Thanks for working on it!

**Comment by rwjblue** [comment]
> in fact I would like to deprecate Route's didTransition action

Not sure I agree here. While I do much prefer to encourage usage of `routeDidChange` over action usage, it is actually pretty annoying to write the code for an individual route to subscribe to `routeDidChange` and determine if it was the target of the transition or if it was above or below the "pivot point" ([this code](https://github.com/nickiaconis/ember-prefetch/blob/master/addon/-private/diff-route-info.js) is the kind of thing I am referring to here).  Until we have a good answer here, I don't feel good about deprecating the actions...

**Comment by xg-wang** [comment]
Forgot to mention when I said I would like to deprecate `didTransition`, a similar `routeDidChange` which bubbles similar to the action event needs be added.

But I think you are right at a better solution may be to have something to  determine if it was the target of the transition or if it was above or below the "pivot point".

**Review by rwjblue — APPROVED** [review]
Thank you!

**Pull request merged**